targets/csharp-unityv2/source/Shared/Internal/PlayFabHttp/PlayFabHTTP.cs

```diff
@@ -0,0 +1,337 @@
+#if !UNITY_WSA && !UNITY_WP8
+            if (PlayFabSettings.RequestType == WebRequestType.HttpWebRequest)
+                _internalHttp = new PlayFabWebRequest();
+#endif
+            if (_internalHttp == null)
+                _internalHttp = new PlayFabWWW();
+            _internalHttp.Awake();
+        }
+
+        /// <summary>
+        /// Internal method for Make API Calls
+        /// </summary>
+        /// <typeparam name="TRequestType"></typeparam>
+        /// <typeparam name="TResultType"></typeparam>
+        /// <param name="api"></param>
+        /// <param name="apiEndpoint"></param>
+        /// <param name="request"></param>
+        /// <param name="authType"></param>
+        /// <param name="resultCallback"></param>
+        /// <param name="errorCallback"></param>
+        /// <param name="customData"></param>
+        protected internal static void MakeApiCall<TRequestType, TResultType>(string api, string apiEndpoint,
+            TRequestType request, string authType, Action<TResultType> resultCallback,
+            Action<PlayFabError> errorCallback, object customData = null)
+        {
+            InitializeHttp();
+            SendEvent(request, null, ApiProcessingEventType.Pre);
+            _internalHttp.MakeApiCall(api, apiEndpoint, request, authType, resultCallback, errorCallback, customData);
+        }
+
+        /// <summary>
+        /// Monobehaviour Awake Method, gathers system data
+        /// </summary>
+        public override void Awake()
+        {
+            _gatherer.GatherData();
+            base.Awake();
+        }
+
+        /// <summary>
+        /// Monobehaviour OnEnable Method, registers logger
+        /// </summary>
+        public void OnEnable()
+        {
+            StartCoroutine(RegisterLogger());
+        }
+
+        /// <summary>
+        /// Monobehaviour OnDisable method, performs cleanup of Logger and bindings to Log Messages
+        /// </summary>
+        public void OnDisable()
+        {
+            if (!string.IsNullOrEmpty(PlayFabSettings.LoggerHost))
+            {
+                _logger = null;
+#if UNITY_5
+                UnityEngine.Application.logMessageReceivedThreaded -= HandleLogOutput;
+#else
+                UniytEngine.Application.RegisterLogCallback(null);
+#endif
+            }
+        }
+
+        /// <summary>
+        /// Monobehaviour Update method, logs System info data if gathered.
+        /// </summary>
+        public void Update()
+        {
+            _internalHttp.Update();
+            if (!_systemInfoLogged && PlayFabSettings.EnableRealTimeLogging)
+            {
+                _gatherer.EnqueueToLogger(_logMessageQueue);
+                _systemInfoLogged = true;
+            }
+        }
+
+        /// <summary>
+        /// Monobehaviour FixedUpdate, logs messages when enabled.
+        /// </summary>
+        void FixedUpdate()
+        {
+
+            lock (_logMessageLock)
```

Comment:
_logMessageQueue is an object, you don't need a separate object to lock.
You have too much work inside of your lock.
You can't do work while you're locked.  You have to transfer data to a local variable, release the lock and THEN operate on the data.